File: src/main/java/com/jcloisterzone/game/capability/DonkeyCapability.java
package com.jcloisterzone.game.capability;

import com.jcloisterzone.action.DonkeyAction;
import com.jcloisterzone.board.PlacementOption;
import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.Rotation;
import com.jcloisterzone.board.Tile;
import com.jcloisterzone.board.pointer.BoardPointer;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.feature.Completable;
import com.jcloisterzone.feature.Feature;
import com.jcloisterzone.feature.River;
import com.jcloisterzone.figure.neutral.Donkey;
import com.jcloisterzone.figure.neutral.Witch;
import com.jcloisterzone.game.Capability;
import com.jcloisterzone.game.Rule;
import com.jcloisterzone.game.capability.trait.FeatureCompletionBlocker;
import com.jcloisterzone.game.state.ActionsState;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.random.RandomGenerator;
import com.jcloisterzone.reducers.MoveNeutralFigure;
import com.jcloisterzone.reducers.PlaceTile;
import io.vavr.collection.List;
import io.vavr.collection.Set;
import io.vavr.collection.Stream;
import io.vavr.Tuple2;

public class DonkeyCapability extends Capability<Void> implements FeatureCompletionBlocker {

    private static final long serialVersionUID = 1L;

    @Override
    public GameState onStartGame(GameState state, RandomGenerator random) {
        state = state.setNeutralFigures(
            state.getNeutralFigures()
                .setDonkey(new Donkey("donkey.1"))
        );
        state = (
            new MoveNeutralFigure<>(state.getNeutralFigures().getDonkey(), new Position(0,0))
        ).apply(state);

        return state;
    }

    @Override
    public boolean isTilePlacementAllowed(GameState state, Tile tile, PlacementOption placement) {
        BoardPointer donkeyPtr = state.getNeutralFigures().getDonkeyDeployment();

        if (donkeyPtr == null) {
        	return true;
        }
        
        Position pos = placement.getPosition();
        Rotation rot = placement.getRotation();

        Integer completedBefore = getCompletedTileFeatures(state, donkeyPtr.getPosition());
        
        state = (new PlaceTile(tile, pos, rot)).apply(state);
        
        Integer completedAfter = getCompletedTileFeatures(state, donkeyPtr.getPosition());

        return completedBefore==completedAfter;
    }

    @Override
    public GameState onActionPhaseEntered(GameState state) {
        ActionsState actions = state.getPlayerActions();

        Donkey donkey = state.getNeutralFigures().getDonkey();

        Set<Position> positions = state.getPlacedTiles().keySet();

        BoardPointer donkeyPtr = state.getNeutralFigures().getDonkeyDeployment();

        if (donkeyPtr != null) {
            positions = positions.filter(fp -> !fp.getPosition().equals(donkeyPtr.getPosition()));
        }
        
        if (donkey != null && positions.size()>0) {
            state = state.appendAction(new DonkeyAction(donkey.getId(), positions));
        }
        return state;
    }
    
    private Integer getCompletedTileFeatures(GameState state, Position pos) {
             
        Integer completed = 0;
        
        for (Tuple2<FeaturePointer, Completable> t : state.getTileFeatures2(pos, Completable.class)) {
        	if (t._2.isCompleted(state)) {
        		if (!(t._2 instanceof River) || state.getBooleanRule(Rule.FISHERMEN)) {
        			completed++;
        		}
            }
        }
        return completed;
    }
    
	public boolean isFeatureCompletionBlocked(GameState state, FeaturePointer fp) {
		Feature f = state.getFeature(fp);

		if (f instanceof Completable c && c.isCompleted(state) && (!(f instanceof River) || state.getBooleanRule(Rule.FISHERMEN))) {
			return c.getTilePositions().contains(state.getNeutralFigures().getDonkeyDeployment().getPosition());
		}
		return false;
	}
}

File: src/main/ts/com/jcloisterzone/game/capability/DonkeyCapability.ts
import { HashSet, type Set } from "../../../../io/vavr/Set.js";
import type { BoardPointer } from "../../board/pointer/BoardPointer.js";
import type { FeaturePointer } from "../../board/pointer/FeaturePointer.js";
import type { PlacementOption } from "../../board/PlacementOption.js";
import { Position } from "../../board/Position.js";
import type { Tile } from "../../board/Tile.js";
import { City } from "../../feature/City.js";
import { type Completable, isInstanceOfCompletable } from "../../feature/Completable.js";
import { Garden } from "../../feature/Garden.js";
import { Monastery } from "../../feature/Monastery.js";
import { River } from "../../feature/River.js";
import { Road } from "../../feature/Road.js";
import { Donkey } from "../../figure/neutral/Donkey.js";
import type { NeutralFigure } from "../../figure/neutral/NeutralFigure.js";
import type { RandomGenerator } from "../../random/RandomGenerator.js";
import { MoveNeutralFigure } from "../../reducers/MoveNeutralFigure.js";
import { PlaceTile } from "../../reducers/PlaceTile.js";
import { DonkeyAction } from "../../action/DonkeyAction.js";
import type { PlayerAction } from "../../action/PlayerAction.js";
import { Capability } from "../Capability.js";
import type { GameState } from "../state/GameState.js";


/** Donkey (Village Life) — a neutral figure that blocks completion of the feature(s)
 *  it sits on: a tile placement that would complete such a feature is forbidden. */
export class DonkeyCapability extends Capability<void> {
  override onStartGame(state: GameState, _random: RandomGenerator): GameState {
    state = state.setNeutralFigures(state.getNeutralFigures().setDonkey(new Donkey("donkey.1")));
    return new MoveNeutralFigure(
      state.getNeutralFigures().getDonkey() as unknown as NeutralFigure<BoardPointer>,
      new Position(0, 0),
    ).apply(state);
  }

  override isTilePlacementAllowed(state: GameState, tile: Tile, placement: PlacementOption): boolean {
    const donkeyPtr = state.getNeutralFigures().getDonkeyDeployment();
    if (donkeyPtr === null) return true;
    const completedBefore = this.getCompletedTileFeatures(state, donkeyPtr);
    const placed = new PlaceTile(tile, placement.getPosition(), placement.getRotation()).apply(state);
    const completedAfter = this.getCompletedTileFeatures(placed, donkeyPtr);
    return completedBefore === completedAfter;
  }

  override onActionPhaseEntered(state: GameState): GameState {
    const donkey = state.getNeutralFigures().getDonkey();
    if (donkey === null) return state;
    const donkeyPos = state.getNeutralFigures().getDonkeyDeployment();
    let positions: Set<Position> = HashSet.ofAll(state.getPlacedTiles().keySet());
    if (donkeyPos !== null) positions = positions.filter((p) => !p.equals(donkeyPos));
    if (positions.isEmpty()) return state;
    return state.appendAction(new DonkeyAction(donkey.getId(), positions) as unknown as PlayerAction<unknown>);
  }

  /** FeatureCompletionBlocker: a completed feature the donkey sits on is blocked
   *  (e.g. a road completed by a tunnel join doesn't score while the donkey is on it). */
  isFeatureCompletionBlocked(state: GameState, fp: FeaturePointer): boolean {
    const f = state.getFeature(fp);
    if (f === null || !isInstanceOfCompletable(f) || !(f as unknown as Completable).isCompleted(state)) return false;
    const fishermen = Capability.tryClassForName("Fishermen");
    const hasFishermen = fishermen !== null && state.hasCapability(fishermen);
    if (f instanceof River && !hasFishermen) return false;
    const donkeyPos = state.getNeutralFigures().getDonkeyDeployment();
    return donkeyPos !== null && (f as unknown as Completable).getTilePositions().contains(donkeyPos);
  }

  /** Number of completed features on the donkey's tile (river only counts with Fishermen). */
  private getCompletedTileFeatures(state: GameState, pos: Position): number {
    const fishermen = Capability.tryClassForName("Fishermen");
    const hasFishermen = fishermen !== null && state.hasCapability(fishermen);
    let completed = 0;
    for (const t of state.getTileFeatures2(pos)) {
      const f = t._2;
      if (!isInstanceOfCompletable(f) || !(f as unknown as Completable).isCompleted(state)) continue;
      if (!(f instanceof River) || hasFishermen) completed++;
    }
    return completed;
  }
}

Capability.register(DonkeyCapability);
